icu4j/main/classes/core/src/com/ibm/icu/text/PluralRules.java

```diff
@@ -2302,6 +2325,31 @@ public double getUniqueKeywordValue(String keyword) {
      * @stable ICU 4.8
      */
     public Collection<Double> getAllKeywordValues(String keyword) {
+        Collection<DecimalQuantity> samples = getAllKeywordDecimalQuantityValues(keyword);
+        if (samples == null) {
+            return null;
+        } else {
+            Collection<Double> result = new LinkedHashSet<>();
+            for (DecimalQuantity dq : samples) {
+                result.add(dq.toDouble());
+            }
+            return result;
```

Comment:
Hmm, yeah. [Java 8 Stream](https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html) is a lot like [Rust Iterator](https://doc.rust-lang.org/std/iter/trait.Iterator.html), with some basic set of functional methods defined in both interfaces.  I suspect that a lot of code here could be expressed as transformations on a sequence, like `.map` (type converting each `DecimalQuantity` to a `double`), `.filter` (removing `double` samples if the type conversion lossiness results in a collision), and that would also be nicely decoupled from the caller choosing which type of concrete collection to materialize to.   Once we are able to verify the upgrade to Java 8 for ICU4J, that might become something to consider.

FWIW, a caveat is that I don't know the performance impacts of using Java 8 Stream.  IIRC, BreakIterator doesn't implement Iterator because ICU was built and integrated with Java before Java 1.2, which when the Iterator interface was introduced. The existing BreakIterator API without Iterator was kept as it was since it worked and callers would probably be slowed down if they had to go through an interface that didn't add much value over what exists.  This caveat may not be all that important here because this code for plural rule sample parsing is primarily consumed by tools and/or internal code.

Reply:
I wasn't suggesting any change, just more expressing my discontent with the state of iterators in Java since Rust does such a good job with them...